invenio_search/engine.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015-2018 CERN.
+# Copyright (C)      2022 University Münster.
+# Copyright (C)      2022 TU Wien.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Transparency module for importing the chosen search engine.
+
+This module imports the ``elasticsearch``/``opensearchpy`` and
+``elasticsearch_dsl``/``opensearch_dsl`` packages based on availability,
+and provides them as ``search`` and ``dsl`` respectively.
+The aim of this is to make the choice of the search engine (Elasticsearch
+vs. OpenSearch) more transparent.
+"""
+
+ES = "Elasticsearch"
+OS = "OpenSearch"
+
+
+try:
+    import elasticsearch as search
+    import elasticsearch_dsl as dsl
+
+    SearchEngine = search.Elasticsearch
+    SEARCH_DISTRIBUTION = ES
+    _fixed_search_version = search.VERSION
+
+except ModuleNotFoundError:
+    import opensearch_dsl as dsl
+    import opensearchpy as search
+
+    SearchEngine = search.OpenSearch
+    SEARCH_DISTRIBUTION = OS
+    _fixed_search_version = search.VERSION
+
+    # FIXME:
+    # to make the transition as transparent as possible for now,
+    # we make OS 1.x look like ES 7.x (of which it is a fork anyway).
+    # for later versions of OS, this should be reworked properly!
+    # NOTE:
+    # `invenio_cli.engine.search.VERSION` will still have the original value
+    if search.VERSION[0] == 1:
+        _fixed_search_version = (7, *search.VERSION[1:])
+
+
+def check_search_version(distribution, version):
+    """Check if the search in use matches the given distribution and version.
+
+    The ``version`` argument can either be a number or a function accepting one
+    argument.
+    In the first case, the specified number will be compared for equality with
+    the major part of the search version (e.g. 7 for ES 7.x).
+    For the second variant, the specified function will be called with the major
+    version as argument. This allows for more elaborate and custom comparisons.
+    The function is expected to return a boolean value.
+    An example would be: ``lambda v: v >= 7``
+    """
+    if SEARCH_DISTRIBUTION.lower() != distribution.lower():
+        return False
+
+    if not callable(version):
+        return search.VERSION[0] == version
+    else:
+        return version(search.VERSION[0])
+
+
+def check_es_version(version):
+    """Convenience alias for ``check_search_version(ES, version)``."""
+    return check_search_version(ES, version)
+
+
+def check_os_version(version):
+    """Convenience alias for ``check_search_version(OS, version)``."""
+    return check_search_version(OS, version)
+
+
+def uses_es7():
```

Comment:
we the check above, this is not needed any more? devs can use `SEARCH_DISTRIBUTION`.

Reply:
i would say that the `uses_es7()` function is independent of the check mentioned above, as it is intended to check for functionality... because OSv1 is basically a fork of ESv7, they provide roughly the same set of features

this is checked in some other modules, e.g. here: https://github.com/inveniosoftware/invenio-records-rest/pull/318/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R266-R267

the right to exist for this function is mostly because `uses_es7()` is just much shorter to write than checking for valid combinations of `SEARCH_DISTRIBUTION` and version numbers (ES>=7 *or* OS==1)

though maybe we should consider renaming this function to something that reflects that more :thinking: